cedar-policy-validator/src/custom_schema/to_json_schema.rs

```diff
@@ -0,0 +1,542 @@
+                match &node.kind {
+                    // Resolve things in the builtin `__cedar` namespace, this includes primitives and extension functions
+                    PathKind::CedarBuiltin(Some(
+                        PrimOrExtension::Decimal | PrimOrExtension::IpAddr,
+                    )) => Some(NamedType::Extension(node.base.node.clone().to_smolstr())),
+                    PathKind::CedarBuiltin(Some(PrimOrExtension::Prim(prim))) => {
+                        Some(NamedType::Primitive(*prim))
+                    }
+                    // Anything else in the builtin `__cedar` namespace does not exist, so fail
+                    PathKind::CedarBuiltin(None) => None,
+                    // Next check for things in the unqualified namespace.
+                    // By default, this includes primitive types and extension functions,
+                    // but can be shadowed by entities and common types
+                    PathKind::Unqualified(Some(prim_or_ext)) => {
+                        // Check if shadowed by common type or entity types
+                        if let Some(typ) =
+                            self.lookup_common_or_entity(&name, node, &common_types, &entity_types)
+                        {
+                            Some(typ)
+                        } else {
+                            // Not shadowed, return the builtin type
+                            match prim_or_ext {
+                                PrimOrExtension::Decimal | PrimOrExtension::IpAddr => {
+                                    Some(NamedType::Extension(node.base.node.clone().to_smolstr()))
+                                }
+                                PrimOrExtension::Prim(prim) => Some(NamedType::Primitive(*prim)),
+                            }
+                        }
+                    }
+                    PathKind::Other | PathKind::Unqualified(None) => {
+                        self.lookup_common_or_entity(&name, node, &common_types, &entity_types)
+                    }
+                }
+            }),
+            collisions,
+        )
+    }
+
+    /// Extract the type corresponding to [`name`] if it exists in the common types or entity types
+    fn lookup_common_or_entity(
+        &self,
+        name: &Node<Name>,
+        node: &Path,
+        common_types: &HashSet<Node<Name>>,
+        entity_types: &HashSet<Node<Name>>,
+    ) -> Option<NamedType> {
+        if common_types.contains(name) || self.global_common_types.contains(name) {
+            Some(NamedType::Common(node.clone().into()))
+        } else if entity_types.contains(name) || self.global_entity_types.contains(name) {
+            Some(NamedType::Entity(node.clone().into()))
+        } else {
+            None
+        }
+    }
+}
+
+pub type TypeNameCollisions = Vec<TypeNameCollision>;
+
+#[derive(Debug, Clone)]
+pub enum TypeNameCollision {
+    CommonTypeAndEntityType(Str, Str),
+    Builtin(Str),
```

Comment:
doc strings?  In particular, what's the `Builtin` -- a `Str` colliding with itself?  (I assume a `Str` colliding with an implied builtin, but a doc string would clarify)